prov/efa/src/efa_cq.c

```diff
@@ -0,0 +1,535 @@
+		return FI_RECV | FI_MSG;
+	case IBV_WC_RDMA_WRITE:
+		return FI_RMA | FI_WRITE;
+	case IBV_WC_RECV_RDMA_WITH_IMM:
+		return FI_REMOTE_CQ_DATA | FI_RMA | FI_REMOTE_WRITE;
+	case IBV_WC_RDMA_READ:
+		return FI_RMA | FI_READ;
+	default:
+		assert(0);
+		return 0;
+	}
+}
+
+static void efa_cq_read_context_entry(struct ibv_cq_ex *ibv_cqx, void *buf)
+{
+	struct fi_cq_entry *entry = buf;
+
+	entry->op_context = (void *)ibv_cqx->wr_id;
+}
+
+static void efa_cq_read_msg_entry(struct ibv_cq_ex *ibv_cqx, void *buf)
+{
+	struct fi_cq_msg_entry *entry = buf;
+
+	entry->op_context = (void *)(uintptr_t)ibv_cqx->wr_id;
+	entry->flags = efa_cq_opcode_to_fi_flags(ibv_wc_read_opcode(ibv_cqx));
+	entry->len = ibv_wc_read_byte_len(ibv_cqx);
+}
+
+static void efa_cq_read_data_entry(struct ibv_cq_ex *ibv_cqx, void *buf)
+{
+	struct fi_cq_data_entry *entry = buf;
+
+	entry->op_context = (void *)ibv_cqx->wr_id;
+	entry->flags = efa_cq_opcode_to_fi_flags(ibv_wc_read_opcode(ibv_cqx));
+	entry->len = ibv_wc_read_byte_len(ibv_cqx);
+	entry->buf = NULL;
+	entry->data = 0;
+
+	if (ibv_wc_read_wc_flags(ibv_cqx) & IBV_WC_WITH_IMM)
+		entry->data = ibv_wc_read_imm_data(ibv_cqx);
+}
+
+static void efa_cq_read_tagged_entry(struct ibv_cq_ex *ibv_cqx, void *buf)
+{
+	struct fi_cq_tagged_entry *entry = buf;
+
+	entry->op_context = (void *)ibv_cqx->wr_id;
+	entry->flags = efa_cq_opcode_to_fi_flags(ibv_wc_read_opcode(ibv_cqx));
+	entry->len = ibv_wc_read_byte_len(ibv_cqx);
+	entry->buf = NULL;
+	entry->data = 0;
+	entry->tag = 0;
+
+	if (ibv_wc_read_wc_flags(ibv_cqx) & IBV_WC_WITH_IMM)
+		entry->data = ibv_wc_read_imm_data(ibv_cqx);
+}
+
+/**
+ * @brief handle the situation that a TX/RX operation encountered error
+ *
+ * This function does the following to handle error:
+ *
+ * 1. write an error cq entry for the operation, if writing
+ *    CQ error entry failed, it will write eq entry.
+ *
+ * 2. increase error counter.
+ *
+ * 3. print warning message with self and peer's raw address
+ *
+ * @param[in]	base_ep
+ * @param[in]	ibv_cq_ex
+ * @param[in]	util_cq     tx/rx cq to write error to
+ * @param[in]	err			positive libfabric error code
+ * @param[in]	prov_errno	positive EFA provider specific error code
+ */
+static void efa_cq_handle_error(struct efa_base_ep *base_ep, struct ibv_cq_ex *ibv_cq_ex, struct util_cq *util_cq, int err, int prov_errno) {
+	struct fi_cq_err_entry err_entry;
+	fi_addr_t addr;
+	int write_cq_err;
+
+	memset(&err_entry, 0, sizeof(err_entry));
+
+	err_entry.err = err;
+	err_entry.prov_errno = prov_errno;
+	err_entry.op_context = (void *)(uintptr_t)ibv_cq_ex->wr_id;
+	err_entry.flags = efa_cq_opcode_to_fi_flags(ibv_wc_read_opcode(ibv_cq_ex));
+	err_entry.buf = NULL;
+	err_entry.data = 0;
+	err_entry.tag = 0;
+
+	if (ibv_wc_read_wc_flags(ibv_cq_ex) & IBV_WC_WITH_IMM)
+		err_entry.data = ibv_wc_read_imm_data(ibv_cq_ex);
+
+	addr = efa_av_reverse_lookup(base_ep->av,
+				     ibv_wc_read_slid(ibv_cq_ex),
+				     ibv_wc_read_src_qp(ibv_cq_ex));
```

Comment:
This is wrong, `ibv_wc_read_src_qp` only applied to the src addr, which means it only applied to the recv call. We need to determine the destination qpn for a send call. If there is an rdma-core API that can does it, we should use it. If it doesn't exist, we need a plan to how to get such addr.

Reply:
I can't find one API for this. Can we save the msg->addr to be part of fi_context?

Reply:
we can, but that doesn't work for inject calls where you don't have a fi_context.